interfaces/builtin/posix_mq.go

```diff
@@ -0,0 +1,237 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/seccomp"
+	"github.com/snapcore/snapd/snap"
+)
+
+const posixMQSummary = `allows access to POSIX message queues`
+
+const posixMQBaseDeclarationSlots = `
+  posix-mq:
+    allow-installation:
+      slot-snap-type:
+        - app
+        - core
+    deny-auto-connection: false
+`
+
+const posixMQPermanentSlotSecComp = `
+mq_open
+mq_getsetattr
+mq_unlink
+mq_notify
+mq_timedreceive
+mq_timedsend
+`
+
+var posixMQPlugPermissions = []string{
+	"open",
+	"associate", /* alias for open */
+	"read",
+	"receive", /* alias for read */
+	"write",
+	"send", /* alias for write */
+	"create",
+	"delete",
+	"destroy", /* alias for delete */
+	"setattr",
+	"getattr",
+}
+
+var posixMQDefaultPlugPermissions = []string{
+	"read",
+	"write",
+}
+
+// Ensure that the name matches the criteria from the mq_overview man page:
+//   Each message queue is identified by a name of the form /somename;
+//   that is, a null-terminated string of up to NAME_MAX (i.e., 255)
+//   characters consisting of an initial slash, followed by one or more
+//   characters, none of which are slashes.
+var posixMQNamePattern = regexp.MustCompile(`^/[^/]{1,255}$`)
+
+type posixMQInterface struct {
+	commonInterface
+}
+
+func (iface *posixMQInterface) StaticInfo() interfaces.StaticInfo {
+	return interfaces.StaticInfo{
+		Summary:           posixMQSummary,
+		ImplicitOnCore:    true,
```

Comment:
Does it make sense for `ImplicitOnCore` or `ImplicitOnClassic`? My understanding is we would always want a matching slot for a given plug and so there would be no implicit version of this interface.

Reply:
Removed